README.md
# jQuery Sticky Table Headers
> Make your table headers sticky

This is a version of the sticky table code originally written by Terry Munn, wrapped in a jquery plugin to make it more reusable. I also added some new options and functions.

* [Example](http://jpchip.github.io/stickytable/example/)
* [Example ES Module](http://jpchip.github.io/stickytable/example/module.html)

## Installing / Getting started

Install via NPM:

```npm
npm install --save jquery-stickytable 
```

Include jQuery(>=1.12.4) and jquery-stickytable in your head:

```html
<script src="path/to/jquery.js"></script>
<script src="path/to/jquery-stickytable.min.js"></script>
```

and the css:

```html
<link rel="stylesheet" type="text/css" href="path/to/jquery-stickytable.css">
```

Files can be found in dist folder:

```text
dist/
├── jquery-stickytable.css
├── jquery-stickytable.js        (UMD)
├── jquery-stickytable.min.js    (UMD, compressed)
└── jquery-stickytable.esm.js    (ES Module, includes css)
```

## Usage

Load stickyTable on your table:

```javascript
$('#myTable').stickyTable(options);
```

By default just the header of the table (eg. the thead) is sticky. Biaxial headers are 
possible — this is for situations where both horizontal and vertical headers are needed. 
To enable, just wrap the first element in each tbody row in a th instead of a td:

```html
  <tbody>
    <tr>
      <th>Sticky column</th>
      <td>value A</td>
      <td>value B</td>
      ...
    </tr>
    ...
  </tbody>
```

## Options

| Option  | Description |
| ------------- | ------------- |
| copyTableClass  | When true copies any classes on the target table to the sticky table. Default is true.  |
| copyEvents  | When true copies any events on the target table to the sticky table. An example would be if there was a click event on a th element in the thead to do sorting, this would carry over to the sticky column header th. Default is false. |
| overflowy | When true adds an overflow-y class to the sticky table wrapper (ie. sticky-wrap class). This means the table will scroll vertically and only be as tall as it's parent div (see .sticky-wrap.overflow-y in css). Default is false.|

## Methods

Methods cannot be called until the stickyTable plugin has been initialized on the element.

### destroy

Completely removes sticky table classes and elements and destroys the instance.

```javascript
  $('#myTable').stickyTable('destroy');
```

## Styles

The sticky headers have default colors set, but you can override this css:

```bash
.sticky-wrap .sticky-intersect th {
    background-color: #666;
    color: #eee;
}

.sticky-wrap th {
	background-color: #666;
	color: #eee;
}
```

## Exceptions

* If target element is not a table.
* If plugin has not been initialised when calling a method
* If method is private
* If method does not exist

## Source

* http://tympanus.net/codrops/2014/01/09/sticky-table-headers-columns/
* http://tympanus.net/Tutorials/StickyTableHeaders/


CHANGELOG.md
Changelog
=========

## Version 3.0.0 - May 8 2019

* Fix script to support newer versions of jquery
* Bundle script to dist folder, create esm version
* Increase debounce time
* Update readme and examples

## Version 2.0.0 - June 6 2017

* Published to NPM
* Moved changelog to own file
* Updated readme and example

## Version 1.1.2 - January 19 2015

* Updated jquery version dependency in bower.json

## Version 1.1.1 - December 31 2014

* Added changelog to readme
* Added version number in comment of css
* Removed dates from code comments
* Added bower.json
* Added example index.html file

## Version 1.1.0 - December 30 2014

* Removed bouncing effect on header and column when scrolling as it was generally annoying.

## Version 1.0 - October 22 2014

* First release


rollup.config.js
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import postcss from 'rollup-plugin-postcss';
import pkg from './package.json';

export default [
	// browser-friendly UMD build
	{
		input: 'src/main.js',
		output: {
			name: 'jquery-stickytable',
			file: pkg.browser,
			format: 'umd'
		},
		context: 'window',
		plugins: [
			resolve(), // so Rollup can find `ms`
			commonjs(), // so Rollup can convert `ms` to an ES module
			postcss({
				extract: true,
				extensions: [ '.css' ],
			})
		]
	},
	{
		input: 'src/main.js',
		external: ['ms'],
		context: 'window',
		output: [
			{ file: pkg.module, format: 'es' }
		],
		plugins: [
			postcss({
				extract: false,
				extensions: [ '.css' ],
			})
		]
	}
];


jquery-stickytable.js
/*
 * jQuery Stick Table Headers - v2.0.0
 * https://github.com/jpchip/stickytable
 * 
 * Wrapped version of sticky table headers code here:
 * http://tympanus.net/codrops/2014/01/09/sticky-table-headers-columns/
 * http://tympanus.net/Tutorials/StickyTableHeaders/
 * 
 * CSS located in jquery-stickytable.css
 * 
 * Depends:
 * jQuery
 * jQuery throttle / debounce - v1.1 - 3/7/2010
 * http://benalman.com/projects/jquery-throttle-debounce-plugin/
 * 
 */
(function ($, window, document, undefined) {
	'use strict';

	var pluginName = 'stickyTable',
	defaults = {
		copyTableClass: true, //copies table classes to sticky tables
		copyEvents: false, //copy events on headers and cols to sticky tables
		overflowy: false //if true limits height of table to height of parent element   
	};

	function StickyTable(element, options) {
		this.element = element;
		this._defaults = defaults;
		this._name = pluginName;

		this.options = $.extend({}, defaults, options);

		this._init();
	}

	StickyTable.prototype._init = function () {
		var $w = $(window),
			$t = $(this.element),
			classList = [],
			$thead,
			$col,
			$stickyHead,
			$stickyCol,
			$stickyInsct,
			$stickyWrap,
			setWidths,
			repositionStickyHead,
			repositionStickyCol,
			calcAllowance;

		if (!$(this.element).is("table")) {
			$.error('Target element must be a table');
		}

		//don't run this function again on same element!
		if ($(this.element).hasClass("sticky-enabled")) {
			return;
		}

		if ($t.find('thead').length > 0 && $t.find('th').length > 0) {
			// Clone <thead>
			$thead = $t.find('thead').clone(this.options.copyEvents);
			$col = $t.find('thead, tbody').clone(this.options.copyEvents);

			// Add class, remove margins, reset width and wrap table
			$t.css({
				margin: 0,
				width: '100%'
			}).wrap('<div class="sticky-wrap" />');

			if (this.options.overflowy) {
				$t.parent().addClass('overflow-y');
			}

			if ($t.attr('class')) {
				classList = $t.attr('class').split(/\s+/);
			}

			$t.addClass('sticky-enabled');

			// Create new sticky table head (basic)
			$t.after('<table class="sticky-thead" />');

			// If <tbody> contains <th>, then we create sticky column and intersect (advanced)
			if ($t.find('tbody th').length > 0) {
				$t.after('<table class="sticky-col" /><table class="sticky-intersect" />');
			}

			// Create shorthand for things
			$stickyHead = $t.siblings('.sticky-thead');
			$stickyCol = $t.siblings('.sticky-col');
			$stickyInsct = $t.siblings('.sticky-intersect');
			$stickyWrap = $t.parent('.sticky-wrap');

			if (this.options.copyTableClass) {
				$.each(classList, function (index, item) {
					$stickyHead.addClass(item);
					$stickyCol.addClass(item);
					$stickyInsct.addClass(item);
				});
			}

			$stickyHead.append($thead);

			$stickyCol.append($col)
				.find('thead th:gt(0)').remove()
				.end()
				.find('tbody td').remove();

			$stickyInsct.html('<thead><tr><th>' + $t.find('thead th:first-child').html() + '</th></tr></thead>');
			$stickyInsct.find('th').height($t.find('thead th:first-child').height());

			setWidths = function () {
				$t.find('thead th').each(function (i) {
					$stickyHead.find('th').eq(i).width($(this).width());
				})
					.end()
					.find('tr').each(function (i) {
						$stickyCol.find('tr').eq(i).height($(this).height());
					});

				// Set width of sticky table head
				$stickyHead.width($t.width());

				// Set width of sticky table col
				$stickyCol.find('th').add($stickyInsct.find('th')).width($t.find('thead th').width());
			};

			repositionStickyHead = function () {
				// Return value of calculated allowance
				var allowance = calcAllowance();

				// Check if wrapper parent is overflowing along the y-axis
				if ($t.height() > $stickyWrap.height()) {
					// If it is overflowing (advanced layout)
					// Position sticky header based on wrapper scrollTop()
					if ($stickyWrap.scrollTop() > 0) {
						// When top of wrapping parent is out of view
						$stickyHead.add($stickyInsct).css({
							opacity: 1,
							top: $stickyWrap.scrollTop()
						});
					} else {
						// When top of wrapping parent is in view
						$stickyHead.add($stickyInsct).css({
							opacity: 0,
							top: 0
						});
					}
				} else {
					// If it is not overflowing (basic layout)
					// Position sticky header based on viewport scrollTop
					if ($w.scrollTop() > $t.offset().top && $w.scrollTop() < $t.offset().top + $t.outerHeight() - allowance) {
						// When top of viewport is in the table itself
						$stickyHead.add($stickyInsct).css({
							opacity: 1,
							top: $w.scrollTop() - $t.offset().top
						});
					} else {
						// When top of viewport is above or below table
						$stickyHead.add($stickyInsct).css({
							opacity: 0,
							top: 0
						});
					}
				}
			};

			repositionStickyCol = function () {
				if ($stickyWrap.scrollLeft() > 0) {
					// When left of wrapping parent is out of view
					$stickyCol.add($stickyInsct).css({
						opacity: 1,
						left: $stickyWrap.scrollLeft()
					});
				} else {
					// When left of wrapping parent is in view
					$stickyCol.css({
						opacity: 0
					})
						.add($stickyInsct).css({
							left: 0
						});
				}
			};

			calcAllowance = function () {
				var a = 0;
				// Calculate allowance
				$t.find('tbody tr:lt(3)').each(function () {
					a += $t.height();
				});

				// Set fail safe limit (last three row might be too tall)
				// Set arbitrary limit at 0.25 of viewport height, or you can use an arbitrary pixel value
				if (a > $w.height() * 0.25) {
					a = $w.height() * 0.25;
				}

				// Add the height of sticky header
				a += $stickyHead.height();
				return a;
			};

			setWidths();

			$t.parent('.sticky-wrap').scroll(function () {
				repositionStickyHead();
				repositionStickyCol();
			});

			$w.load(setWidths)
				.resize($.debounce(250, function () {
					setWidths();
					repositionStickyHead();
					repositionStickyCol();
				}))
				.scroll(repositionStickyHead);
		}
	};

	StickyTable.prototype.destroy = function () {
		var $t = $(this.element),
			$stickyHead = $t.siblings('.sticky-thead'),
			$stickyCol = $t.siblings('.sticky-col'),
			$stickyInsct = $t.siblings('.sticky-intersect');

		this.options = $.extend({}, defaults);

		if ($t.hasClass('sticky-enabled')) {
			$stickyHead.remove();
			$stickyCol.remove();
			$stickyInsct.remove();

			$t.unwrap();
			$t.removeClass('sticky-enabled');
			//todo: clear events...
		}
		$.removeData(this.element, 'plugin_' + this._name);
	};

	$.fn[pluginName] = function (options) {
		return this.each(function () {
			var instance = $.data(this, 'plugin_' + pluginName);
			if (instance) {
				if (typeof options === 'string' || options instanceof String) {
					if (instance[options] !== undefined && $.isFunction(instance[options])) {
						if (options.indexOf('_') !== -1) {
							$.error('Method ' + options + ' is private!');
						} else {
							instance[options](Array.prototype.slice.call(arguments, 1));
						}
					} else {
						$.error('Method ' + options + ' does not exist.');
					}
				}
			} else if (typeof options === 'object' || !options) {
				$.data(this, 'plugin_' + pluginName, new StickyTable(this, options));
			} else {
				$.error('Plugin must be initialised before using method: ' + options);
			}
		});
	};

}(jQuery, window, document));


src/main.js
/**
 * jQuery Stick Table Headers - v3.0.0
 * https://github.com/jpchip/stickytable
 *
 * Wrapped version of sticky table headers code here:
 * http://tympanus.net/codrops/2014/01/09/sticky-table-headers-columns/
 * http://tympanus.net/Tutorials/StickyTableHeaders/
 *
 * CSS located in jquery-stickytable.css
 *
 * Depends:
 * jQuery
 * jQuery throttle / debounce - v1.1 - 3/7/2010
 * http://benalman.com/projects/jquery-throttle-debounce-plugin/
 * @preserve
 */

import './jquery-stickytable.css';
import './jquery.ba-throttle-debounce.min';

(function ($, window, document, undefined) {
	'use strict';

	var pluginName = 'stickyTable',
	defaults = {
		copyTableClass: true, //copies table classes to sticky tables
		copyEvents: false, //copy events on headers and cols to sticky tables
		overflowy: false //if true limits height of table to height of parent element
	};

	function StickyTable(element, options) {
		this.element = element;
		this._defaults = defaults;
		this._name = pluginName;

		this.options = $.extend({}, defaults, options);

		this._init();
	}

	StickyTable.prototype._init = function () {
		var $w = $(window),
			$t = $(this.element),
			classList = [],
			$thead,
			$col,
			$stickyHead,
			$stickyCol,
			$stickyInsct,
			$stickyWrap,
			setWidths,
			repositionStickyHead,
			repositionStickyCol,
			calcAllowance;

		if (!$(this.element).is("table")) {
			$.error('Target element must be a table');
		}

		//don't run this function again on same element!
		if ($(this.element).hasClass("sticky-enabled")) {
			return;
		}

		if ($t.find('thead').length > 0 && $t.find('th').length > 0) {
			// Clone <thead>
			$thead = $t.find('thead').clone(this.options.copyEvents);
			$col = $t.find('thead, tbody').clone(this.options.copyEvents);

			// Add class, remove margins, reset width and wrap table
			$t.css({
				margin: 0,
				width: '100%'
			}).wrap('<div class="sticky-wrap" />');

			if (this.options.overflowy) {
				$t.parent().addClass('overflow-y');
			}

			if ($t.attr('class')) {
				classList = $t.attr('class').split(/\s+/);
			}

			$t.addClass('sticky-enabled');

			// Create new sticky table head (basic)
			$t.after('<table class="sticky-thead" />');

			// If <tbody> contains <th>, then we create sticky column and intersect (advanced)
			if ($t.find('tbody th').length > 0) {
				$t.after('<table class="sticky-col" /><table class="sticky-intersect" />');
			}

			// Create shorthand for things
			$stickyHead = $t.siblings('.sticky-thead');
			$stickyCol = $t.siblings('.sticky-col');
			$stickyInsct = $t.siblings('.sticky-intersect');
			$stickyWrap = $t.parent('.sticky-wrap');

			if (this.options.copyTableClass) {
				$.each(classList, function (index, item) {
					$stickyHead.addClass(item);
					$stickyCol.addClass(item);
					$stickyInsct.addClass(item);
				});
			}

			$stickyHead.append($thead);

			$stickyCol.append($col)
				.find('thead th:gt(0)').remove()
				.end()
				.find('tbody td').remove();

			$stickyInsct.html('<thead><tr><th>' + $t.find('thead th:first-child').html() + '</th></tr></thead>');
			$stickyInsct.find('th').height($t.find('thead th:first-child').height());

			setWidths = function () {
				$t.find('thead th').each(function (i) {
					$stickyHead.find('th').eq(i).width($(this).width());
				})
					.end()
					.find('tr').each(function (i) {
						$stickyCol.find('tr').eq(i).height($(this).height());
					});

				// Set width of sticky table head
				$stickyHead.width($t.width());

				// Set width of sticky table col
				$stickyCol.find('th').add($stickyInsct.find('th')).width($t.find('thead th').width());
			};

			repositionStickyHead = function () {
				// Return value of calculated allowance
				var allowance = calcAllowance();

				// Check if wrapper parent is overflowing along the y-axis
				if ($t.height() > $stickyWrap.height()) {
					// If it is overflowing (advanced layout)
					// Position sticky header based on wrapper scrollTop()
					if ($stickyWrap.scrollTop() > 0) {
						// When top of wrapping parent is out of view
						$stickyHead.add($stickyInsct).css({
							opacity: 1,
							top: $stickyWrap.scrollTop()
						});
					} else {
						// When top of wrapping parent is in view
						$stickyHead.add($stickyInsct).css({
							opacity: 0,
							top: 0
						});
					}
				} else {
					// If it is not overflowing (basic layout)
					// Position sticky header based on viewport scrollTop
					if ($w.scrollTop() > $t.offset().top && $w.scrollTop() < $t.offset().top + $t.outerHeight() - allowance) {
						// When top of viewport is in the table itself
						$stickyHead.add($stickyInsct).css({
							opacity: 1,
							top: $w.scrollTop() - $t.offset().top
						});
					} else {
						// When top of viewport is above or below table
						$stickyHead.add($stickyInsct).css({
							opacity: 0,
							top: 0
						});
					}
				}
			};

			repositionStickyCol = function () {
				if ($stickyWrap.scrollLeft() > 0) {
					// When left of wrapping parent is out of view
					$stickyCol.add($stickyInsct).css({
						opacity: 1,
						left: $stickyWrap.scrollLeft()
					});
				} else {
					// When left of wrapping parent is in view
					$stickyCol.css({
						opacity: 0
					})
						.add($stickyInsct).css({
							left: 0
						});
				}
			};

			calcAllowance = function () {
				var a = 0;
				// Calculate allowance
				$t.find('tbody tr:lt(3)').each(function () {
					a += $t.height();
				});

				// Set fail safe limit (last three row might be too tall)
				// Set arbitrary limit at 0.25 of viewport height, or you can use an arbitrary pixel value
				if (a > $w.height() * 0.25) {
					a = $w.height() * 0.25;
				}

				// Add the height of sticky header
				a += $stickyHead.height();
				return a;
			};

			setWidths();

			$t.parent('.sticky-wrap').scroll(function () {
				repositionStickyHead();
				repositionStickyCol();
			});

			$w.on('load', setWidths)
				.resize($.debounce(500, function () {
					setWidths();
					repositionStickyHead();
					repositionStickyCol();
				}))
				.scroll(repositionStickyHead);
		}
	};

	StickyTable.prototype.destroy = function () {
		var $t = $(this.element),
			$stickyHead = $t.siblings('.sticky-thead'),
			$stickyCol = $t.siblings('.sticky-col'),
			$stickyInsct = $t.siblings('.sticky-intersect');

		this.options = $.extend({}, defaults);

		if ($t.hasClass('sticky-enabled')) {
			$stickyHead.remove();
			$stickyCol.remove();
			$stickyInsct.remove();

			$t.unwrap();
			$t.removeClass('sticky-enabled');
			//todo: clear events...
		}
		$.removeData(this.element, 'plugin_' + this._name);
	};

	$.fn[pluginName] = function (options) {
		return this.each(function () {
			var instance = $.data(this, 'plugin_' + pluginName);
			if (instance) {
				if (typeof options === 'string' || options instanceof String) {
					if (instance[options] !== undefined && $.isFunction(instance[options])) {
						if (options.indexOf('_') !== -1) {
							$.error('Method ' + options + ' is private!');
						} else {
							instance[options](Array.prototype.slice.call(arguments, 1));
						}
					} else {
						$.error('Method ' + options + ' does not exist.');
					}
				}
			} else if (typeof options === 'object' || !options) {
				$.data(this, 'plugin_' + pluginName, new StickyTable(this, options));
			} else {
				$.error('Plugin must be initialised before using method: ' + options);
			}
		});
	};

}(jQuery, window, document));
